packages/CodeQuality/src/Rector/If_/ShortenElseIfRector.php

```diff
@@ -0,0 +1,100 @@
+<?php declare(strict_types=1);
+
+namespace Rector\CodeQuality\Rector\If_;
+
+use Nette\Utils\Strings;
+use PhpParser\Node;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Expr\Assign;
+use PhpParser\Node\Expr\Ternary;
+use PhpParser\Node\Stmt;
+use PhpParser\Node\Stmt\Else_;
+use PhpParser\Node\Stmt\ElseIf_;
+use PhpParser\Node\Stmt\Expression;
+use PhpParser\Node\Stmt\If_;
+use Rector\Rector\AbstractRector;
+use Rector\RectorDefinition\CodeSample;
+use Rector\RectorDefinition\RectorDefinition;
+
+/**
+ * @see \Rector\CodeQuality\Tests\Rector\If_\ShortenElseIfRector\ShortenElseIfRectorTest
+ */
+final class ShortenElseIfRector extends AbstractRector
+{
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition('Shortens else/if to elseif', [
+            new CodeSample(
+                <<<'PHP'
+class SomeClass
+{
+    public function run()
+    {
+        if ($cond1) {
+            return $action1;
+        } else {
+            if ($cond2) {
+                return $action2;
+            }
+        }
+    }
+}
+PHP
+                ,
+                <<<'PHP'
+class SomeClass
+{
+    public function run()
+    {
+        if ($cond1) {
+            return $action1;
+        } elseif ($cond2) {
+            return $action2;
+        }    
+    }
+}
+PHP
+            ),
+        ]);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [If_::class];
+    }
+
+    /**
+     * @param If_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (!$else = $node->else) {
+            return null;
+        }
+
+        if (count($else->stmts) > 1) {
+            return null;
+        }
+
+        $if = $else->stmts[0];
+        if (! $if instanceof If_) {
+            return null;
+        }
+
+        $if = $this->refactor($if) ?? $if;
```

Comment:
This is pretty difficult to understand recursion.
Could you think of more readable syntax?

Reply:
Removed the coalesc and made it more verbose, 
hopefully that makes it clearer

Do you have any better ideas in mind how to clarify this?

Reply:
I'm not sure why have you chosen this approach, so I don't even understand the idea to describe it

Reply:
Ok the reason behind the recursion is that we get rid of the original nested `If_` node, but this rector runs only on `If_` nodes.
This means after refactoring the top-level `If_` we won't detect any more else/if condition that we could refactor inside that if statement, because now it is an `Elseif_` Node.
Consider the example from the `recursive` fixture:
```php
    public function run()
    {
        if ($this->cond1) {
            $this->doSomething();
        } else {
            if ($this->cond2) {
                $this->doSomething();
            } else {
                if ($this->cond3) {
                    $this->doSomething();
                }
            }
        }
    }
```
In that case we would move the `if ($this->cond2)` statement up and combine it to the `elseif ($this->cond2)`. By this refactoring our rector doesn't run for the nested `if ($this->cond3)`, it would get refactored if we run the rector a second time.
Therefore we have to run the refactoring on the inner `if ($this->cond3)` before applying the refactoring on the outer `if ($this->cond3)`.

I hope this clarifies the need for recursion, i'll try to express this inside the code

Reply:
I see 

> I hope this clarifies the need for recursion, i'll try to express this inside the code

That would be perfect, so this comment is not needed to understand the mechanism